test-schema.json

```diff
@@ -1,5 +1,6 @@
 {
     "$schema": "https://json-schema.org/draft/2020-12/schema",
+    "$id": "https://raw.githubusercontent.com/json-schema-org/JSON-Schema-Test-Suite/main/test-schema.json",
```

Comment:
Why `githubusercontent.com` instead of `json-schema.org`?

Reply:
Not sure what you mean. We don't have it hosted there. If you mean why don't we add it, see the PR description, where yes maybe that'd be nice, but I figured we'd wait till the new site was up to decide where or how that goes given it's not terribly pressing to put it there to me (nothing else is asking for it at the minute).

Reply:
@Julian oh I see.  I was just thinking that it's not actually necessary to have it _hosted_ on json-schema.org immediately, so we might as well use the canonical URI that we'd keep.  But I guess then there's the question of directory layout?

It's just that if people start referencing it with this `$id` then that kind of has to work forever without changing. And as I understand it, that URI will always reflect the current file on main, which will change particularly as we add output testing and the like.

Reply:
> But I guess then there's the question of directory layout?

Yes, who knows what that URI would be, we don't have the new site yet.

> It's just that if people start referencing it with this $id then that kind of has to work forever without changing.

What do you mean by having to work forever without changing? That the contents don't change? Isn't this essentially the same as the old unversioned metaschema URI (https://json-schema.org/schema#)?

Reply:
> What do you mean by having to work forever without changing? That the contents don't change? Isn't this essentially the same as the old unversioned metaschema URI (https://json-schema.org/schema#)?

Yes, and we got rid of that because it was a problem that the contents changed.  That URI now returns an HTML page explaining that it should not be used, with `meta refresh` redirect (which means that the client has to actually follow the redirect, it won't transparently just receive the schema).

We used to do bugfix updates in place for meta-schemas but even that was problematic which is why the spec now has this langauge:

> [8.1.3. ](https://www.ietf.org/archive/id/draft-bhutton-json-schema-01.html#section-8.1.3)[Updates to Meta-Schema and Vocabulary URIs](https://www.ietf.org/archive/id/draft-bhutton-json-schema-01.html#name-updates-to-meta-schema-and-)
Updated vocabulary and meta-schema URIs MAY be published between specification drafts in order to correct errors. Implementations SHOULD consider URIs dated after this specification draft and before the next to indicate the same syntax and semantics as those listed here.

which allows us to deploy a 2022-08 meta-schema if we want to fix a bug that we notice this month (not that we've done that, but I added that section because various spec team members objected to updating in place and I had to admit they had a point).

Reply:
That doesn't seem like an issue here honestly. The spec changes often and replaces itself, so it was an issue that the whole world's schemas all suddenly shift versions when we release a new versions. Here we have a relatively static or generally append-only thing where our downstream users anyhow don't have a current versioning expectation. But will have another look tomorrow.

Reply:
The issue is whether the schema can be downloaded and cached indefinitely, in order to prevent people from fetching it all the time.  The spec recommends "long-lived" cache headers for this reason.

> Here we have a relatively static or generally append-only thing where our downstream users anyhow don't have a current versioning expectation. But will have another look tomorrow.

Given the number of major expansions needed within the tests suite, I'm skeptical of this claim.  We need to add annotation and output support (and output itself is a format that is likely to change).  We need to add error case support.  We need to add tests for how extension vocabularies are supported.  There's also the the stuff that notEthan just submitted.  Why would the test format be expected to stay stable given all of that?

Reply:
To clarify:   I'm not so much trying to block this PR as I am just trying to figure out what the collective policy is here, because I thought we had a pretty solid consensus on not changing schemas at a given URI.  Mostly because I used to be in favor of it and got convinced otherwise.

I hope that eventually we do come to and document a consensus on this for all official JSON Schema Organization schemas, including the test suite, and the test suite schema(s) end up primarily hosted on the new web site.  But in the meantime, while I would not go with this `$id` I'm not blocking the PR and assuming someone else approves it I don't need anything addressed before it is merged.

Reply:
> [EDIT: I realized the simple answer is you were probably planning on those being different files]

Yes, exactly.

As for never changing, in practice 99% of the people using this schema get it alongside the tests, and "caching" effectively is until the next time they pull. The reality is anyone fetching it over the network is undoubtedly *already* fetching it from that URI, so the intention here was just to effectively preserve the status quo. I suspect there's consensus amongst us that schemas which change over time are bad effectively because they represent APIs that change over time without notice -- but that's something true here *without* this URI, so it's that part that's bad :), but it for better or worse has worked for us. Nevertheless I realized after sending this that it may not be needed at all to use the schema with schemastore, so that's perhaps a reason regardless to punt on the PR until the new site is up -- or I could perhaps just use a URI for a tagged version or something, which I guess also avoids the issue as well but creates some annoyance in remembering to update it.

(I'm not rushing to merge this PR given it's not important for schemastore now I think, so I'll leave it for a bit and see whether anyone else chimes in before deciding what to do with it).